### PR TITLE
bpo-31425: Expose AF_QIPCRTR in socket module

### PR DESCRIPTION
**Pull request by andersson**
The AF_QIPCRTR address family was introduced in Linux v4.7, expose this
in the Python socket module to allow applications to operate on sockets
of this type.

Signed-off-by: Bjorn Andersson <bjorn.andersson@linaro.org>

https://bugs.python.org/issue31425

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by andersson** [comment]
The CLA is signed, but unfortunately I did not have the "Github name" filled in on my account. This is corrected now.

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -2127,6 +2159,14 @@ getsockaddrlen(PySocketSockObject *s, socklen_t *len_ret)
     }
 #endif
 
+#if defined(AF_QIPCRTR)
+    case AF_QIPCRTR:
+    {
+        *len_ret = sizeof (struct sockaddr_qrtr);
```
Comment (tiran) [inline Modules/socketmodule.c:2303]: please remove the extra space after sizeof

File: Modules/socketmodule.h [hunk 2]
```diff
@@ -54,6 +54,15 @@ typedef int socklen_t;
 #  undef AF_NETLINK
 #endif
 
+#ifdef HAVE_LINUX_QRTR_H
+# ifdef HAVE_ASM_TYPES_H
+#  include <asm/types.h>
```
Comment (tiran) [inline Modules/socketmodule.h:59]: You must add the new addr type to the large ``sock_addr_t`` union.

**Review by tiran — CHANGES_REQUESTED** [review]
Please add tests and maybe an example, too.

File: Modules/socketmodule.c (hunk 1 above)
Comment (andersson) [reply Modules/socketmodule.c:2303]: For the record, I'm ignoring this to follow the coding style in this function.

**Review by tiran — APPROVED** [review]
PR looks fine to me

@vstinner @serhiy-storchaka Can you have a look, please? We don't have a dedicated expert for the socket module. You have touched some code recently, too.

File: Modules/socketmodule.c [hunk 3]
```diff
@@ -7,8 +7,8 @@ This module provides an interface to Berkeley socket IPC.
 Limitations:
 
 - Only AF_INET, AF_INET6 and AF_UNIX address families are supported in a
-  portable manner, though AF_PACKET, AF_NETLINK and AF_TIPC are supported
-  under Linux.
+  portable manner, though AF_PACKET, AF_NETLINK, AF_QIPCRTR and AF_TIPC are
```
Comment (serhiy-storchaka) [inline Modules/socketmodule.c:10]: Explain the address for this family below as it is explained for other families.

File: Modules/socketmodule.c [hunk 4]
```diff
@@ -1579,6 +1587,30 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
     }
 #endif /* AF_NETLINK */
 
+#if defined(AF_QIPCRTR)
+    case AF_QIPCRTR:
+    {
+        struct sockaddr_qrtr* addr;
+        int node, port;
```
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]: `unsigned int`

File: Modules/socketmodule.c [hunk 5]
```diff
@@ -1579,6 +1587,30 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
...
+        }
+        if (!PyArg_ParseTuple(args, "II:getsockaddrarg", &node, &port))
+            return 0;
+        addr->sq_family = AF_QIPCRTR;
+        addr->sq_node = node;
+        addr->sq_port = port;
+        *len_ret = sizeof(*addr);
+        return 1;
+    }
+#endif
```
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]: Add a comment as for other `#endif`s.

File: Modules/socketmodule.c [hunk 6]
```diff
@@ -2125,6 +2157,14 @@ getsockaddrlen(PySocketSockObject *s, socklen_t *len_ret)
     }
 #endif /* AF_NETLINK */
 
+#if defined(AF_QIPCRTR)
+    case AF_QIPCRTR:
+    {
+        *len_ret = sizeof (struct sockaddr_qrtr);
+        return 1;
+    }
+#endif
```
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]: Add a comment as for other `#endif`s.

File: Doc/library/socket.rst [hunk 7]
```diff
@@ -161,6 +161,13 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.7
 
+- :const:`AF_QIPCRTR` is a Linux-only socket based interface for communicating
+  with services running on co-processors in Qualcomm platforms. The address
+  family is represented as a ``(node, port)`` tuple where the node and port are
```
Comment (serhiy-storchaka) [inline Doc/library/socket.rst]: `the *node* and *port*`

File: Doc/library/socket.rst [hunk 8]
```diff
@@ -161,6 +161,13 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.7
 
+- :const:`AF_QIPCRTR` is a Linux-only socket based interface for communicating
```
Comment (serhiy-storchaka) [inline Doc/library/socket.rst:196]: Document `AF_QIPCRTR` below in the Constants section.

**Comment by tiran** [comment]
@andersson I'm sorry that your PR didn't make it into Python 3.7.

Please rebase your PR on master and address @serhiy-storchaka's comments.

**Comment by csabella** [comment]
@andersson   Any chance you would be able to respond to @tiran and @serhiy-storchaka 's requests? Seems like this is close to being able to be merged.  Thanks!

**Comment by taleinat** [comment]
Due to this being nearly ready to go but untouched for over 6 months, I've gone ahead and addressed @serhiy-storchaka's latest review comments, and merged the latest master branch in.

@serhiy-storchaka, @tiran, consider reviewing again.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you @taleinat! All my comments have been addressed.

**Comment by andersson** [comment]
Thanks for picking this up @taleinat and thanks for merging it @serhiy-storchaka!

File: Doc/library/socket.rst [hunk 9]
```diff
@@ -193,6 +193,13 @@ created.  Socket addresses are represented as follows:
   - *addr* - Optional bytes-like object specifying the hardware physical
     address, whose interpretation depends on the device.
 
+- :const:`AF_QIPCRTR` is a Linux-only socket based interface for communicating
+  with services running on co-processors in Qualcomm platforms. The address
+  family is represented as a ``(node, port)`` tuple where the *node* and *port*
+  are non-negative integers.
+
+  .. versionadded:: 3.7
```
Comment (ZackerySpytz) [inline Doc/library/socket.rst:201]: This should be 3.8, not 3.7.

Comment (taleinat) [reply Doc/library/socket.rst:201]: Good catch!

**Pull request merged**